src/main/java/de/unijena/cheminf/mortar/controller/PipelineSettingsViewController.java

```diff
@@ -205,46 +215,48 @@ private void addListenerAndBindings() {
      *                                pipeline settings view is constructed for initial display and true when the user adds
-     *                                a new row using the "+" button
+     *                                a new row using the "+" button; only used when first param is not null
      */
     private void addNewChoiceRow(String aFragmenterDisplayName, boolean anUpdateFragmentersList) {
         ComboBox<String> tmpComboBox = new ComboBox<>();
-        for (IMoleculeFragmenter tmpFragmenter : this.fragmenters) {
+        for (IMoleculeFragmenter tmpFragmenter : this.availableFragmentersMasterArray) {
             tmpComboBox.getItems().add(tmpFragmenter.getFragmentationAlgorithmDisplayName());
         }
+        //remains if given fragmenter name is null
         tmpComboBox.setPromptText(Message.get("PipelineSettingsView.comboBox.promptText"));
         tmpComboBox.setOnAction(anActionEvent -> {
-            Object tmpSelectedFragmenterString = tmpComboBox.getSelectionModel().getSelectedItem();
+            String tmpSelectedFragmenterDisplayName = tmpComboBox.getSelectionModel().getSelectedItem();
             int tmpIndex = GridPane.getRowIndex(tmpComboBox) - 1;
-            for (IMoleculeFragmenter tmpFragmenter : this.fragmenters) {
-                if (tmpSelectedFragmenterString.equals(tmpFragmenter.getFragmentationAlgorithmDisplayName())) {
+            for (IMoleculeFragmenter tmpFragmenter : this.availableFragmentersMasterArray) {
+                if (tmpSelectedFragmenterDisplayName.equals(tmpFragmenter.getFragmentationAlgorithmDisplayName())) {
                     // will not work cause size of list is set - why not, it is in case a fragmenter that has already been set is changed
-                    if (this.fragmenterList.size() > tmpIndex) {
-                        this.fragmenterList.set(tmpIndex, Arrays.stream(this.fragmenters)
-                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterString)).findFirst().orElse(null));
+                    if (this.selectedPipelineFragmentersList.size() > tmpIndex) {
+                        this.selectedPipelineFragmentersList.set(tmpIndex, Arrays.stream(this.availableFragmentersMasterArray)
+                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterDisplayName)).findFirst().get().copy());
                     } else {
-                        this.fragmenterList.add(Arrays.stream(this.fragmenters)
-                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterString)).findFirst().orElse(null));
+                        this.selectedPipelineFragmentersList.add(Arrays.stream(this.availableFragmentersMasterArray)
+                                .filter(x -> x.getFragmentationAlgorithmDisplayName().equals(tmpSelectedFragmenterDisplayName)).findFirst().get().copy());
```

Comment:
@FelixBaensch might I ask what is the reason behind this stream filtering here? We already identified the fragmenter to use via the for loop and the "if name equals fragmenter.name" condition, right?

Reply:
And there is also another "actual" change here, the added copy() calls.

Reply:
I have double checked this, and If I see it correctly this is your code :)

Reply:
Irc, this is legacy code, as we only allowed each fragmenter once per pipeline